### PR TITLE
[dynamo] Initial exception handling support

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #126304
* #126578
* #126956
* __->__ #126923

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

### CODE REVIEW

File: torch/_dynamo/variables/builtin.py [hunk 1]
```diff
@@ -712,6 +712,20 @@ def _make_handler(fn, arg_types: List[type], has_kwargs: bool):
                 tx, [v.realize() for v in args], kwargs
             )
 
+        if inspect.isclass(fn) and issubclass(fn, Exception):
+
+            def create_exception_class_object(tx, args, kwargs):
+                if fn is AssertionError and not all(
```
Comment (anijain2305) [inline torch/_dynamo/variables/builtin.py:718]: This is just to maintain the status quo for test `pytest test/dynamo/test_repros.py::ReproTests::test_rewrite_assert_with_non_string_msg`

File: test/dynamo/test_repros.py [hunk 2]
```diff
@@ -3051,7 +3051,7 @@ def f(x):
         with self.assertRaisesRegex(AssertionError, "torch.Size"):
             opt_f(args)
         self.assertEqual(
-            torch._dynamo.utils.counters["unimplemented"][
+            torch._dynamo.utils.counters["graph_break"][
```
Comment (anijain2305) [inline test/dynamo/test_repros.py:3054]: Earlier it was a SkipFrame, now its a graph break.

**Review by williamwen42 — APPROVED** [review]

File: torch/_dynamo/exc.py [hunk 3]
```diff
@@ -183,6 +183,10 @@ class IncorrectUsage(Exception):
     pass
 
 
+class ObservedException(TorchDynamoException):
```
Comment (jansel) [inline torch/_dynamo/exc.py:186]: comment explaing what this is

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**